### ISSUE TITLE
Problems with running the Unity environment

### ISSUE THREAD

**Issue by chenjq20** [Ameyapores/SafeRLSurgery#1]
Hi,

I am very interested in the manipulation of soft tissue but I have met some problems about running the Unity environment. 

I have installed Unity ML Agent Toolkit (release 11) in Unity and some key python packages in the virtual environment. The versions of the packages are as followed:

`mlagents    0.23.0`
`mlagents-envs   0.23.0`
`torch   1.7.0+cu110`
`tensorboard 2.8.0`

When I added NNModel to agent (behavior 1-10) and press the "Play" button in Unity, the PSM2 would not move. In the debug console message, there were two messsages looping:

`GraspFat active(double click here for see flag definition): False`
`relgrasp (double click here for see flag definition): False`

I also tried to run the environment with the command "mlagents-learn". While I have not assigned the config file yet, this command could run and output results, but with this following output:

`WARNING [trainer_factory.py:68] BehaviorPPO does not match any behaviors specified in the trainer configuration file: []`
`UnityEnvironment worker 0: environment stopping`
`mlagents_envs.exception.UnityTimeOutException: The Unity environment took too long to respond. Make sure that:`
`The environment does not need user interaction to launch`
`The Agents' Behavior Parameters > Behavior Type is set to "Default"`
`The environment and the Python interface have compatible versions`

I have tried to test the model in the builded env with running "main.py". I have ran into some problems with this code:

`from gym_unity.envs import UnityEnv`

I installed gym-unity with the version of 0.23.0, but "UnityEnv" could not be found.

I wonder the proper way to play with this environment. Thank you for any help you can offer!

**Comment by Ameyapores** [comment]
Hi,
My apologies, I haven't updated the repository (I will do it in the next couple of days).
I used mlagents==0.16.0.
Could you test the mlagents-learn with this version?

The debug log messages can be ignored. They are a flag when the instrument grasps the soft tissue.

**Comment by chenjq20** [comment]
Hi,

Thank you for your kind help!

I have tested the mlagents-learn with mlagents==0.16.0. The version information with python environment is as followed:

`Version information:`
`ml-agents:0.16.0`
`ml-agents-envs:0.16.0`
`Communicator API:1.0.0`
`TensorFlow:2.8.0`

When I kept the Unity environment unchanged and open the project, the original package information shown in Package Manager is as followed:

`Barracuda preview-1.2.1-preview`
`ML Agents preview-1.7.0-preview`

I manually changed the com.unity.ml-agents version to 1.0.0 with [ML-Agents Release 1](https://github.com/Unity-Technologies/ml-agents/releases/tag/release_1), several errors occurred and the play mode could not executed. These errors did not exist with com.unity.ml-agents version 1.7.0.

`..\ml-agents-release_1\ml-agents-release_1\com.unity.ml-agents\Runtime\Agent.cs(156,30): error CS0246: The type or namespace name 'BehaviorParameters' could not be found (are you missing a using directive or an assembly reference?)`
`..\ml-agents-release_1\ml-agents-release_1\com.unity.ml-agents\Runtime\Inference\BarracudaModelParamLoader.cs(132,49): error CS0246: The type or namespace name 'BehaviorType' could not be found (are you missing a using directive or an assembly reference?)`
`..\ml-agents-release_1\ml-agents-release_1\com.unity.ml-agents\Runtime\Agent.cs(160,9): error CS0246: The type or namespace name 'BehaviorParameters' could not be found (are you missing a using directive or an assembly reference?)`
`..\ml-agents-release_1\ml-agents-release_1\com.unity.ml-agents\Runtime\Inference\BarracudaModelParamLoader.cs(132,77): error CS0103: The name 'BehaviorType' does not exist in the current context`

I wonder how I could set the Unity environment properly. I have kept the Barracuda version unchanged. Should I install another version of Barracuda which matches the ML agents version? Or should I change the way of install com.unity.ml-agents? (I change the version by "Add package from disk".)

Also, I wonder which config file I should use with mlagents-learn. I currently use train_config.yaml in ML-Agents Release 1.

Thank you again for any help you can offer!

**Comment by Ameyapores** [comment]
Hi,

I found some time to clone the repo and to open the scene. Indeed, I am not able to move the instrument.
The following configurations seem to give no error:

Unity version: 2019.4.13f1 or 2019.4.19f1
Barracuda preview-1.2.1-preview
ML Agents preview-1.7.0-preview

However, I couldn't figure out why EEpoint2 is not moving based on OnActionReceived() function in Agentscript. I will investigate on this part.
For the config file, you can use the default one. However, in case you use gym_unity, you can make your own agent/DRL architecture. Can you give it a try with a lower version of gym_unity: 0.21.0.

I will give you an update tomorrow with mlagents and hopefully get the gym_unity pipeline working as well.
Meanwhile, we have another project for soft tissue manipulation: https://gitlab.com/altairLab/unityflexml
It is simpler however its a bit outdated but seems to work

**Comment by chenjq20** [comment]
Hi,

Thank you very much for your help!

I have tried with gym_unity: 0.21.0 (with ml-agents and ml-agents-envs:0.23.0). However, the same error still existed with this lower version.

`ImportError: cannot import name 'UnityEnv' from 'gym_unity.envs'`

I have also tried with UnityFlexML and it works with behavior model. But I am confused with the training pipeline in UnityFlexML. Meanwhile, I am also interested in how Unity C# SDK is linked to the Python API. It will be very helpful if this SafeRLSurgery environment works well.

Thank you again for your kind help!

**Comment by Ameyapores** [comment]
Hello @chenjq20 ,

I managed to clean all the unnecessary files and scripts from the project. Also, I upgraded the mlagents package version to 1.0.8 which can be installed from the package manager of Unity, with Barracuda 1.0.4. The scene in the latest commit has it installed. I tested it with mlagents-learn config/trainer_config.yaml with mlagents=0.16.1 installed in my conda environment, and the training works. But I think you can use higher versions as well, as long as the config file is fine.

I changed the main.py file for gym_unity. Check main2.py where 
`gym_unity.envs import UnityToGymWrapper` . They changed the name from UnityEnv to UnityToGymWrapper.

`gym_unity : 0.26.0` 
`mlagents-envs : 0.26.0` 
`tensorflow: 2.8.0`

Although the brain (neural network) loaded is not valid, it doesn't move the instrument. But the Python API communication to Unity C# SDK works.

**Comment by chenjq20** [comment]
Hi,

Thank you a lot for your kind help!

I tested with two version of environment, but something went wrong with both training pipeline. In the first version of environment, I got the version exactly same as pre-requisite versions:
 
`ml-agents: 0.16.1`
`ml-agents-envs: 0.26.0`
`Communicator API: 1.5.0`
`TensorFlow:2.8.0`
`gym-unity:0.26.0`

I kept the Unity environment unchanged with mlagents 1.0.8 and  Barracuda 1.0.4. My Unity version is Unity 2019.4.19f1.

When I tested mlagents-learn with config/trainer_config.yaml in release 2，I got this warning:

`W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudart54_110.ddl'; dlerrror: cudar64_110,dll not found`

Then learning was interrupted by a TypeError:

`File "D:\anaconda3\envs\ml-unity\lib\site-packages\mlagents\trainer\learn.py", line 485, in create_envrionment`
`side_channels=side_channels,`
`Type_Error:__init__( ) got an unexpected keyword argument 'args'`

When I tested main2.py with this version of environment, the built environment seemed to have some problems with load part of the scene. Only the instrument , the fat and the kidney were shown in the scene. And the running process got the same warning and the following info:

`W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudart54_110.ddl'; dlerrror: cudar64_110,dll not found`
`[INFO] Connected to Unity environment with package version 1.0.8 and communication version 1.0.0`

The training process could not go on and would time out.

Then I tested a higher version of mlagents:0.26.0 (installed from the packages release 17), which required torch to run. I got the torch version of 1.10.2+cu113.

With this environment, I tested mlagents-learn with config/ppo/Basic.yaml in release 17. It could run with "action1" in Unity Console but the instrument would not move.

The warnings and output remained the same with main2.py.

 I wonder if the problem is related with tensorflow version or inference device configuration (cpu or gpu). But I have not idea about the lost parts of scene in the built environment. I will try other tensorflow versions and inference device configurations. I will also try editor option with main2.py.

Thank you again for any help you can offer!

**Comment by chenjq20** [comment]
Hi,

I have another try in the first version of environment (with mlagents 0.16.1). By manually installing 'cudart64_110.ddl' (sorry for the type error in the last comment), the warning about this dll has been fixed. However, the other problems still existed. 

When I tested mlagents-learn with config/trainer_config.yaml in release 2, learning was interrupted by a TypeError:

`File"D:\anaconda3\envs\ml-unity\lib\site-packages\mlagents\trainer\learn.py", line 485, in create_unity_envrionment`
`side_channels=side_channels,`
`Type_Error:__init__( ) got an unexpected keyword argument 'args'`

I built a new scene from the unity environment and replaced the "built_scene" folder and a full scene could be used with main2.py. When I tested main2.py with this version of environment, there were no warning or error, but the instrument in the built environment would not move, and the process ended with:

`The Unity environment took too long to respond...`
`[WARNING] Environment timed out shutting down. Killing...`

I wonder if more user interaction should be made to launch the environment in the test scene and how editor option could be used. 

Thank you very much for any help you can offer!

**Comment by Ameyapores** [comment]
Hi,

`W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudart54_110.ddl'; dlerrror: cudar64_110,dll not found`  This is a warning associated with GPU. It affects the performance significantly when vision based inputs such as images are used for training the agent. In this project, we use physical kinematics such as position of the robot. Hence, this warning does not affect much. But I see that you solved the issue by manually installing 'cudart64_110.ddl'. Unfortunately, there is no good/proper documented way to solve this warning. I did the same.

Here is a step by step guide that worked for me. Before, I was using to separate environment for mlagents and gym_unity. Now I created a single environment with the following package configuration. 
`mlagents: 0.20.0`
`mlagents-envs: 0.20.0`
`gym-unity: 0.20.0`
`gym: 0.20.0`
`tensorflow: 2.3.0`
`tensorboard: 2.7.0`

Then you can use mlagents-learn with config/ppo/Basic.yaml in release 17. I tested with release 15. and start the training from the editor. You can see the instrument moving in the scene and training on the command line.
When you use mlagents-learn with config/trainer_config.yaml in release 2, you will get the following error
`mlagents.trainers.exception.TrainerConfigError: The option default was specified in your YAML file, but is invalid.`
In this case, you need to downgrade to mlagents:0.16.1. and it works fine. But there might be some dependency problems with gym_unity. Hence, i recommend to use the first configuration with release 17.

As for gym_unity, you can specify unity_env = UnityEnvironment(None) instead of unity_env = UnityEnvironment(cfg['train']['name']) to start from the unity editor. Right now the script is only for testing. You will need to create a training script using something similar on this page:  https://github.com/Unity-Technologies/ml-agents/blob/main/gym-unity/README.md

You can ignore the action1 debug log in the console or comment it out in the Agentscript.cs file on line 91.
I had disabled some part of the scene. I built it again in the new commit. Should be fine now.
Let me know if it worked for you :)

**Comment by chenjq20** [comment]
Hi,

Thanks a lot for your help! 

I have just tested the new commit and environment. It worked for me perfectly in both the running ways.  I am sincerely grateful for your kind help.

I also found that the warning concerned about GPU was not solved. Since raw images may be used in the future applications,  I will investigate on this GPU issue. The main2.py test script worked well for me. I will also try creating similar training scripts.

Many thanks for you warm help! Have a nice day:)
